File: CountingSort/countingsort.py
def countingSort(arr):
    if len(arr) == 0:
        return arr
    max = arr[0]
    for j in arr:
        if j > max:
            max = j
    new_arr = [0 for i in range(0, max + 1)]
    for i in range(len(arr)):
        new_arr[arr[i]] += 1

    for k in range(1, len(new_arr)):
        new_arr[k] = new_arr[k] + new_arr[k - 1]

    sorted_arr = [0 for i in range(len(arr))]

    for t in range(len(arr)-1,-1,-1): # to make it more stable
        sorted_arr[new_arr[arr[t]] - 1] = arr[t]
        new_arr[arr[t]] -= 1

    return sorted_arr

tests = [
    ([1, 2, 3, 4, 5], [1, 2, 3, 4, 5]),
    ([5, 4, 3, 2, 1], [1, 2, 3, 4, 5]),
    ([3, 1, 2, 3, 1], [1, 1, 2, 3, 3]),
    ([7], [7]),
    ([], []),
    ([4, 2, 2, 8, 3, 3, 1], [1, 2, 2, 3, 3, 4, 8])
]

for arr, expected in tests:
    result = countingSort(arr[:])
    print(f"Input: {arr}, Output: {result}, Expected: {expected}, OK: {result == expected}")

File: CountingSort/CountingSort.java
package CountingSort;

public class CountingSort {

	public static void countingSort(int [] arr) {
        if(arr.length == 0)
            return;	
            
        int max = arr[0];
        for(int i = 0; i < arr.length; i++){
            if(arr[i] > max)
                max = arr[i];
        }
        int[] counting = new int[max + 1];
        for(int j = 0; j < arr.length; j++){
            counting[arr[j]] += 1;	
            }
        for(int k = 1; k < counting.length; k++){
            counting[k] = counting[k] + counting[k-1];
        }
        int[] output = new int[arr.length];
        for(int l = arr.length - 1; l >= 0; l--){
            output[counting[arr[l]] - 1] = arr[l];
            counting[arr[l]] -= 1;
        }
        System.arraycopy(output, 0, arr, 0, arr.length);
}
public static void main(String[] args) {
        int[][] tests = {
            {1, 4, 1, 2, 7, 5, 2}, 
            {5, 4, 3, 2, 1},         
            {1, 2, 3, 4, 5},         
            {3, 3, 2, 1, 2, 1},      
            {7},                     
            {}                       
        };

        int[][] expected = {
            {1, 1, 2, 2, 4, 5, 7},
            {1, 2, 3, 4, 5},
            {1, 2, 3, 4, 5},
            {1, 1, 2, 2, 3, 3},
            {7},
            {}
        };

        for (int i = 0; i < tests.length; i++) {
            CountingSort.countingSort(tests[i]);
            System.out.print("Test " + (i + 1) + ": Output = ");
            for (int x : tests[i]) System.out.print(x + " ");
            System.out.print("| Expected = ");
            for (int x : expected[i]) System.out.print(x + " ");
            System.out.println();
        }
    }


}
